packages/gateway/src/routes/verifyCode/handler.ts

```diff
@@ -0,0 +1,231 @@
+  nonce: string,
+  code: string
+): Promise<void> {
+  const codeDetails: ICodeDetails = await getVerificationCodeDetails(nonce)
+
+  if (!codeDetails) {
+    throw new Error('sms code not found')
+  }
+
+  const codeExpired =
+    (Date.now() - codeDetails.createdAt) / 1000 >=
+    CONFIG_SMS_CODE_EXPIRY_SECONDS
+
+  if (code !== codeDetails.code) {
+    throw new Error('sms code invalid')
+  }
+
+  if (codeExpired) {
+    throw new Error('sms code expired')
+  }
+}
+
+export async function getVerificationCodeDetails(
+  nonce: string
+): Promise<ICodeDetails> {
+  const codeDetails = await get(`verification_${nonce}`)
+  return JSON.parse(codeDetails) as ICodeDetails
+}
+
+export async function deleteUsedVerificationCode(
+  nonce: string
+): Promise<boolean> {
+  const count = await del(`verification_${nonce}`)
+  return Boolean(count)
+}
+
+export function generateNonce() {
+  return crypto
+    .randomBytes(16)
+    .toString('base64')
+    .toString()
+}
+
+export async function sendVerificationCode(
+  mobile: string,
+  verificationCode: string
+): Promise<void> {
+  const params = {
+    msisdn: mobile,
+    code: verificationCode
+  }
+
+  await fetch(resolve(NOTIFICATION_URL, 'authenticationCode'), {
+    method: 'POST',
+    body: JSON.stringify(params),
+    headers: {
+      Authorization: `Bearer ${await createToken(
+        'auth',
+        ['service'],
+        ['opencrvs:notification-user'],
+        JWT_ISSUER
+      )}`
+    }
+  })
+
+  return undefined
+}
+
+export async function createToken(
+  userId: string,
+  scope: string[],
+  audience: string[],
+  issuer: string,
+  system?: boolean
+): Promise<string> {
+  if (typeof userId === undefined) {
+    throw new Error('Invalid userId found for token creation')
+  }
+  return sign({ scope }, cert, {
+    subject: userId,
+    algorithm: 'RS256',
+    expiresIn: system
+      ? CONFIG_SYSTEM_TOKEN_EXPIRY_SECONDS
+      : CONFIG_TOKEN_EXPIRY_SECONDS,
+    audience,
+    issuer
+  })
+}
+
+export async function generateAndSendVerificationCode(
+  nonce: string,
+  mobile: string,
+  scope: string[]
+) {
+  const isDemoUser = scope.indexOf('demo') > -1
+  logger.info('isDemoUser', {
+    isDemoUser: isDemoUser
+  })
+  let verificationCode
+  if (isDemoUser) {
+    verificationCode = '000000'
+    await storeVerificationCode(nonce, verificationCode)
+  } else {
+    verificationCode = await generateVerificationCode(nonce, mobile)
+  }
+  if (!PRODUCTION || QA_ENV) {
+    logger.info('Sending a verification SMS', {
+      mobile: mobile,
+      verificationCode
+    })
+  } else {
+    if (isDemoUser) {
+      throw unauthorized()
+    } else {
+      await sendVerificationCode(mobile, verificationCode)
+    }
+  }
+}
+
+export default async function sendVerifyCodeHandler(
+  request: Hapi.Request,
+  h: Hapi.ResponseToolkit
+): Promise<ISendVerifyCodeResponse> {
+  const payload = request.payload as ISendVerifyCodePayload
+  const { userId, phoneNumber, scope } = payload
```

Comment:
Can we get userId and scope from the user's verified authorization token instead? Notice that as this currently is, any user could send any other user's id here as it's not validated. A malicious user could also send a scope of ["demo"] with any phone number and know the verification code is `000000`

Reply:
When you do this, then you don't have to send the data anymore either from the frontend, which I think will simplify things there quite a bit

Reply:
Thanks for your excellent suggestion. I have modified the code to take userId and scope from the request header token and removed parameters from both the backend and front-end.